### ISSUE TITLE
cell_data: cell_type->name to name->cell_type?

### ISSUE THREAD

**Issue by nschloe** [nschloe/meshio#615]
I've been working with cell data a little bit recently and got somewhat annoyed by the fact that, right now, meshio organizes the data in a nested dict as, e.g.,
```python
cell_data["triangle"]["density"] = ...
```
It gets a little tedious then to check which cell data items are present in the mesh since you have to go through all cell types, fetch the keys, compare them to the keys in the other cell types etc.

Of course, this organization makes it easy for meshes to define a cell data item on triangles and not quads, but really I have never seen this.

This is why I'm thinking about swapping the dict order to
```python
cell_data["density"]["triangle"] = ...
```
Of course that'd need bumping to 4.* (which I'm not terribly opposed against though).

What do y'all think about that? I want to hear @tianyikillua, @gdmcbain, @clbarnes, @MuellerSeb, @keurfonluu.

**Comment by tianyikillua** [comment]
By doing so, the list of available cell fields would be
```
cell_data_list = list(cell_data.keys())
```
which would be the same for point data 
```
point_data_list = list(point_data.keys())
```
I always say yes to consistency.

**Comment by MuellerSeb** [comment]
I like this suggestion. It's a much more logical order for the mesh data.

**Comment by gdmcbain** [comment]
I didn't have an opinion but the way that @tianyikillua puts it, that `cell_data` and `point_data` will have the same keys, is good.

In the same way that @nschloe has

> been working with cell data a little bit recently

I'm thinking that since I discovered the new `cell_sets`, I'll be using them extensively. They should get the same swap, I suppose; there's less of an issue with breaking existing code there.

**Comment by keurfonluu** [comment]
It makes more sense to me too.

**Comment by clbarnes** [comment]
I haven't interacted with cell data at all, but agree that consistency is good - I defer to you.

The only thing I'd add is wondering whether there's scope for replacing any level of the nested dicts with e.g. a dataclass, if you know the keys ahead of time. Of course we want to be able to represent arbitrary data for any given feature, but if it could be made any more introspectable, I'd be in favour of that. That would also help set the update apart from the previous implementation, rather than just switching the order of two string keys, which is hard to demonstrate.

**Comment by gdmcbain** [comment]
I like dataclasses a lot and use them (and frozen dataclasses) whenever possible, but here (besides not yet being able to count on Python 3.7 for a while) I don't think we do know the keys in advance, not the fields (very application-specific) but also not the cell-types, assuming we don't want to insist on having all of the types (do we have dozens) with most being empty and fix that list of types (more might be desired in future).

**Comment by gdmcbain** [comment]
On incrementing the meshio major version, another big breaking change on the horizon I think is reimplementing  MSH4.1. The initial version was hasty and basically coerced it back to MSH2. I suspect the fix is to use `cell_sets` (instead of `cell_data[cell_type]["gmsh:physical"]`).  I take this to correspond to the main design intent behind Gmsh's update; MSH4.1 implements tagged subsets of elements #290. I think that handling this in meshio is required to fix #524 and #614.

That will break a lot of user code so might need a new major version. Should the present proposal, which meets with general approval but doesn't seem all that urgent, be staged for the same major version release?

**Comment by tianyikillua** [comment]
Yes I also think that the next major version should unify the notion of point / cell sets among formats. Maybe we should continue the discussion in another issue?

In quick, is your `cell_sets` designed as follows?
```
cell_sets = {
     "Top": {"line": [1, 2]},
    "Material A": {"quad": [0, 1, 5], "triangle": [1, 10]},
}
```

**Comment by gdmcbain** [comment]
> In quick, is your cell_sets designed as follows?

Yes, I think so: `Dict[str, Dict[str, ndarray]]`, where the first key depends on the application, the second is the type of cell, and the value is an array of indices into `cells[cell_type ]`.

> Let's pursue the discussion in #619.

Yes. Ta.

**Comment by MuellerSeb** [comment]
**Another thought about that:**

As I have mentioned before in #183 it would be nice, if the cell order could be preserved!
The background of sorting cells by there type was the different number of nodes for different cell types, so they could not be put in one numpy array. So they were separated by their cell-type to have the same number of nodes to fill an 2D numpy array. This was an early decision, but unfortunately this was destroying the cell-order.

I think almost all mesh formats work in the following way:

- define nodes with IDs
- define cells with type and list of Nodes
- define data (cell/node/field...) by IDs

If we could access the cell-ID, we wouldn't need to sort cell-data again by the cell-type. We could rather have one numpy array, where the first dimension corresponds to the cell ID.

The fun thing is, that there is an implicit cell-ID within each cell-type, which is used for the cell-data (since it's an numpy array, where again the first dimension corresponds with the cell-sub-ID from the specified cell-type).

**TL;DR:**
Having one numpy array per cell-data would be the ideal case in my opinion.
Therefore we would need the cell-ID from the mesh file, which is not preserved at the moment.

**Comment by gdmcbain** [comment]
> If we could access the cell-ID, we wouldn't need to sort cell-data again by the cell-type. We could rather have one numpy array, where the first dimension corresponds to the cell ID.

I have a common use-case that I don't see how to fit into that structure. I have triangles for the domain and lines for the boundary; e.g. this is what one gets from Gmsh. Many cell-data fields will only be defined on the triangles, not the lines. Would I have to put a dummy NaN value on the lines?

**Comment by nschloe** [comment]
Thanks everyone for your input.

There are many good ideas here, and I think it's time to go ahead and introduce some breaking changes. The next release will be 4.0, I'm thinking about a month from now. If you have ideas (like the dataclasses or the cell ordering), please open a new issue. Copy-and-pasting from the comments here is completely fine of course.

I've already incremented the version requirement to 3.6, and will reorder the dict as suggested in the original comment here. (That's not to say that'll be the final data model for 4.0.)

**Comment by MuellerSeb** [comment]
> I have a common use-case that I don't see how to fit into that structure. I have triangles for the domain and lines for the boundary; e.g. this is what one gets from Gmsh. Many cell-data fields will only be defined on the triangles, not the lines. Would I have to put a dummy NaN value on the lines?

So you are basically lucky, that the cell-data was divided by cell-type in that case.
What happens, if there are some cells left out in a cell-data component?
How does meshio handle that at the moment? @nschloe

**Comment by nschloe** [comment]
> Many cell-data fields will only be defined on the triangles, not the lines. Would I have to put a dummy NaN value on the lines?

There are different notions of what constitutes a "cell function". It commonly is treated as a _function_ that associates a value to every cell. In meshio, it's theoretically possible to only keep values with a particular cell type,
```python
cell_data = {"a": {"line": [0.5, 1.4, 3.13]}}
```
but this is probably not what you want. I think it makes sense to first think about tags/sets because this is what a "boundary" is.

**Comment by gdmcbain** [comment]
> but this is probably not what you want.

On the contrary.  I would indeed use something like `cell_data["flux"]["line"]` for the flux through the boundary integrated over each element.  This quantity is defined only on the boundary and not in the domain and therefore only on the lines and not on the triangles.  

Similarly I might project a P1 or P2 solution on the domain to be uniform on each cell (‘P0’) and then I'd have `cell_data["potential"]["triangle"]`.  This quantity is only defined in the domain and not on the boundary and therefore on the triangles and not on the lines.

> I think it makes sense to first think about tags/sets because this is what a "boundary" is.

Yes, I do agree with this, but this feels a fair way away in the future.  

My current thoughts are, that, say my domain consists of triangles, then the boundary should be tagged edges of these triangles.  These are called 'facets' in *Wikipedia* [‘Simplicial complex’](https://en.wikipedia.org/wiki/Simplicial_complex) and [scikit-fem](
https://github.com/kinnala/scikit-fem/blob/f8909f5d1f95ec2f83ee5141b1f3ce43b5591894/skfem/mesh/mesh2d/mesh_tri.py#L280-L285) and 'Face' in [*XDMF Model and Format* (‘Attribute’)](http://www.xdmf.org/index.php/XDMF_Model_and_Format#Attribute).  I don't think this part of XDMF is yet supported by meshio:

https://github.com/nschloe/meshio/blob/ab6c1792795c9f9e18d8b9a25a960e6c7b6e007b/meshio/xdmf/main.py#L290-L294

The main reason though why I have `cell_data` defined on either `"triangle"` or `"line"` but not both is that this is what `meshio.read` returns for MSH2.2 Physical Surfaces (subdomains) and Physical Curves (parts of the boundary in a mixed boundary value problem), respectively.  These tags `cell_data["gmsh:physical"]["line"]` and `cell_data["gmsh:physical"]["triangle"]` are only defined on one or the other.  

Sets would be better for both subdomains and parts of the boundary, but I don't have access to an implementation.  My Gmsh is not compiled against MED.

**Comment by keurfonluu** [comment]
I probably missed something but, wouldn't it make more sense for ``cell_data`` to follow the same logic as the new ``cells``? I mean to have ``cell_data[name]`` as a list of ``Cells`` in the same order as ``cells``...

**Comment by tianyikillua** [comment]
You are right. It's been done already. https://github.com/nschloe/meshio/pull/631#issuecomment-576905837

**Comment by nschloe** [comment]
> You are right. It's been done already.

Indeed.
